src/drivers/optical_flow/upflow/lc302.cpp

```diff
@@ -0,0 +1,640 @@
+	if (_fd > -1) {
+		::close(_fd);
+	}
+
+	if (_reports) {
+		delete _reports;
+	}
+
+	if (_upflow_lc302_topic) {
+		orb_unadvertise(_upflow_lc302_topic);
+	}
+
+	perf_free(_sample_perf);
+	perf_free(_collect_timeout_perf);
+	perf_free(_comm_error);
+}
+
+int upflow_lc302::_fd_open()
+{
+	if (!_serial_port) {
+		return -1;
+	}
+
+	_fd = ::open(_serial_port, O_RDWR | O_NOCTTY | O_NONBLOCK);
+
+	if (_fd < 0) {
+		PX4_ERR("open upflow_lc302_serial_port:%s faild",_serial_port);
+		return -1;
+	}
+
+	struct termios config;
+
+	int r = tcgetattr(_fd, &config);
+
+	if (r) {
+		PX4_ERR("Unable to get termios from %s.", _serial_port);
+		goto error;
+	}
+
+	/* clear: data bit size, two stop bits, parity, hardware flow control */
+	config.c_cflag &= ~(CSIZE | CSTOPB | PARENB | CRTSCTS);
+	/* set: 8 data bits, enable receiver, ignore modem status lines */
+	config.c_cflag |= (CS8 | CREAD | CLOCAL);
+	/* turn off output processing */
+	config.c_oflag = 0;
+	/* clear: echo, echo new line, canonical input and extended input */
+	config.c_lflag &= (ECHO | ECHONL | ICANON | IEXTEN);
+
+	r = cfsetispeed(&config, B19200);
+	r |= cfsetospeed(&config, B19200);
+
+	if (r) {
+		PX4_ERR("Unable to set baudrate");
+		goto error;
+	}
+
+	r = tcsetattr(_fd, TCSANOW, &config);
+
+	if (r) {
+		PX4_ERR("Unable to set termios to %s", _serial_port);
+		goto error;
+	}
+
+	return PX4_OK;
+
+error:
+	::close(_fd);
+	_fd = -1;
+	return PX4_ERROR;
+}
+
+int upflow_lc302::init()
+{
+	if (CDev::init()) {
+		PX4_ERR("Unable to initialize device\n");
+		return -1;
+	}else{
+		PX4_INFO("able to initialize device");
+	}
+
+	_reports = new ringbuffer::RingBuffer(2, sizeof(optical_flow_s));
+
+	if (!_reports) {
+		PX4_ERR("No memory to allocate RingBuffer");
+		return -1;
+	}else{
+		PX4_INFO("new ringbuffer success");
+	}
+
+	if (_fd_open()) {
+		PX4_ERR("_fd_open faild");
+		return PX4_ERROR;
+	}else{
+		PX4_INFO("_fd_open success ");
+	}
+
+	hrt_abstime timeout_usec, now;
+
+	for (now = hrt_absolute_time(), timeout_usec = now + PROBE_USEC_TIMEOUT;
+	     now < timeout_usec;
+	     now = hrt_absolute_time()) {
+
+		//PX4_INFO("now:%d",(int)now);
+		if (_cycle() > 0) {
+			printf(NAME " found\n");
+			return PX4_OK;
+		}
+
+		px4_usleep(1000);
+	}
+
+	PX4_ERR("No readings from " NAME);
+
+	return PX4_ERROR;
+}
+
+void
+upflow_lc302::stop()
+{
+	work_cancel(HPWORK, &_work);
+}
+
+void upflow_lc302::start()
+{
+	/*
+	 * file descriptor can only be accessed by the process that opened it
+	 * so closing here and it will be opened from the High priority kernel
+	 * process
+	 */
+	::close(_fd);
+	_fd = -1;
+
+	work_queue(HPWORK, &_work, (worker_t)&upflow_lc302::cycle_trampoline, this, USEC2TICK(WORK_USEC_INTERVAL));
+}
+
+void
+upflow_lc302::cycle_trampoline(void *arg)
+{
+	upflow_lc302 *dev = reinterpret_cast<upflow_lc302 * >(arg);
+
+	if (dev->_fd != -1) {
+		dev->_cycle();
+
+	} else {
+		dev->_fd_open();
+	}
+
+	work_queue(HPWORK, &(dev->_work), (worker_t)&upflow_lc302::cycle_trampoline,
```

Comment:
Needs to be updated to use the new WorkQueue.

Reply:
Yes, but I'm not familiar on this part, I need your help.

Reply:
Tried to fixed.